### DISCUSSION TITLE
Using Either and Bind when the outputs of two separate functions feeds into the next function

### DISCUSSION THREAD

**Discussion by JonTheTurnip** [louthy/language-ext#1210] [Q&A] [answered]
Hi there,

Is there an approach I should use when using Either and Bind, when the outputs of two separate functions feed into the next function?

I could just refactor the whole thing in various ways to get around the problem, but I'm guessing there's a cleaner and less contrived way?

In the example below my DoMapping function needs both a Something and an Input, therefore I can't chain them altogether with Bind.  Hope that makes sense?

```csharp
Either<Error, Input> ValidateInput(Input input) => input;
Either<Error, Something> GetSomething() => new Something();
Either<Error, MappedInput> DoMapping(Input input, Something something) => new MappedInput();
Either<Error, Response> CallSomethingExternal(MappedInput input) => new Response();
```

**Comment by JonTheTurnip** [comment, marked as answer]
I think I've just discovered that if I use the Linq query syntax for select many then you can achieve this very easily?  Still curious to know how best to do this using Bind method chaining though.

e.g

```csharp

var result =
    await (
        from a in GetInputs()
        from b in UserConfirm(a).ToAsync()
        from c in Validate(b).ToAsync()
        from d in GetSomething().ToAsync()
        from e in DoMapping(c, d).ToAsync()
        select e
    );

```

**Reply by louthy** [reply]
```c#
GetInputs().Bind(
    a => UserConfirm(a).ToAsync().Bind(
        b => Validate(b).ToAsync().Bind(
            c => Validate(b).ToAsync().Bind(
                d => GetSomething().ToAsync().Bind(
                    e => DoMapping(c, d).ToAsync())
```
In other words, use LINQ  ;)

**Reply by JonTheTurnip** [reply]
Great, many thanks for confirming!
